runtime/compiler/control/CompilationThread.cpp

```diff
@@ -4417,7 +4417,24 @@ TR::CompilationInfoPerThread::processEntry(TR_MethodToBeCompiled &entry, J9::J9S
       }
    else // compilation will not be retried
       {
-      //fprintf(stderr, "compilation thread will free optimization plan %p %d\n", entry->_optimizationPlan, entry->_optimizationPlan->creatorCanFreePlan());
+#if defined(J9VM_OPT_JITSERVER)
+      if(compInfo->getPersistentInfo()->getRemoteCompilationMode() == JITServer::CLIENT &&
+         !entry.isRemoteCompReq() &&
+         entry._optimizationPlan->getOptLevel() == cold)
+         {
+         // If it's a local cold downgraded compilation, add an upgrade request to the LPQ, hoping that
+         // by the time requests from it start being processed, the server is available
+         if (TR::Options::isAnyVerboseOptionSet(
+                TR_VerboseJITServer,
+                TR_VerboseCompilationDispatch,
+                TR_VerbosePerformance,
+                TR_VerboseCompFailure))
+            TR_VerboseLog::writeLineLocked(TR_Vlog_JITServer, "t=%6u Buffering an upgrade request into the LPQ: j9method=%p",
+                                           (uint32_t) compInfo->getPersistentInfo()->getElapsedTime(),
+                                           entry.getMethodDetails().getMethod());
+         compInfo->getLowPriorityCompQueue().addUpgradeReqToLPQ(&entry);
+         }
+#endif
```

Comment:
Some compilations are performed at cold because we want to: maybe the options forced us to compile at cold or maybe they are simple enough that a recompilation would not improve the code. JNI and thunks are typically compiled at cold and we don't want to upgrade them. Moreover, to overcome network latency, some compilations that were downgraded for other reasons (say because we want to improve start-up) are performed locally. These should not be added to the LPQ. We need to pass some information down, from the point we decide we want to add a compilation to LPQ, to this point where we actually add the request to LPQ. Also, we may need to update the reason that entry exists in LPQ, so that the logic which extracts from LPQs can be changed.